TargetComposition/SymBuilder/ApiProvider.cpp

```diff
@@ -1064,20 +1161,80 @@ Object FieldsObject::Add(_In_ const Object& /*fieldsObject*/,
                          _In_ ComPtr<UdtTypeSymbol>& spUdtTypeSymbol,
                          _In_ std::wstring fieldName,
                          _In_ Object fieldType,
-                         _In_ std::optional<ULONG64> fieldOffset)
+                         _In_ size_t argCount,                        // [fieldOffset], [details]
+                         _In_reads_(argCount) Object *pArgs)
 {
     SymbolSet *pSymbolSet = spUdtTypeSymbol->InternalGetSymbolSet();
 
-    ULONG64 fieldOffsetToUse = fieldOffset.value_or(UdtPositionalSymbol::AutomaticAppendLayout);
-    BaseTypeSymbol *pFieldType = UnboxType(pSymbolSet, fieldType);
+    ULONG64 fieldOffsetToUse = UdtPositionalSymbol::AutomaticAppendLayout;
+
+    //
+    // We are "overloaded" but must do our own overload resolution.  An ordinal is the field offset.  An object
+    // is the details object which supplies bitfield positions, etc...
+    //
+    size_t detailsArg = 0;
+    if (argCount > 0)
+    {
+        if (pArgs[0].GetKind() == ObjectIntrinsic)
+        {
+            detailsArg = 1;
+            fieldOffsetToUse = (ULONG64)(pArgs[0]);
+        }
+    }
+
+    bool isBitField = false;
+    ULONG64 bitFieldPosition = UdtPositionalSymbol::AutomaticAppendLayout;
+    ULONG64 bitFieldLength = 0;
+
+    if (argCount > detailsArg)
+    {
+        Object details = pArgs[detailsArg];
+        std::optional<Object> bitFieldLengthObj = details.TryGetKeyValue(L"BitFieldLength");
+        if (bitFieldLengthObj.has_value())
+        {
+            bitFieldLength = (ULONG64)bitFieldLengthObj.value();
+            isBitField = true;
+        }
+
+        if (isBitField)
+        {
+            std::optional<Object> bitFieldPositionObj = details.TryGetKeyValue(L"BitFieldPosition");
+            if (bitFieldPositionObj.has_value())
+            {
+                bitFieldPosition = (ULONG64)bitFieldPositionObj.value();
+            }
+        }
+    }
+
+    //
+    // Explicitly allow type names such as "int :3" to be unboxed.  The unboxing will return int; however,
+    // UnboxType will return the "3" as a value so that we can apply it.
+    //
+    // NOTE: You cannot specify an explicit 'BitFieldLength' in the details object which conflicts with
+    //       whatever you put in the type name if you are using a string type name.
+    //
+    ULONG64 typeNameBitFieldLength;
+    BaseTypeSymbol *pFieldType = UnboxType(pSymbolSet, fieldType, true, &typeNameBitFieldLength);
+
+    if (typeNameBitFieldLength != 0)
+    {
+        if (bitFieldLength != 0 && bitFieldLength != typeNameBitFieldLength)

```

Comment:
No, I can't.  It's perfectly valid to have typeNameBitFieldLength == 3 (e.g.: you said "int:3"), bitFieldLength == 0 (you didn't specify a manual length in the details object) and that shouldn't throw.  Removing the != 0 there would cause that to suddenly throw an exception.